Bug Fix: allow emitted particles to live out their lives on stopEmit()

andrewleek-droplab opened the pull request:
_This is a redo of this PR: https://github.com/creativelifeform/three-nebula/pull/206 but done from its own branch....(pardon the user error)_

This fixes a bug where emitted particles would stop animating when calling stopEmit() on an emitter. Now they will live out the rest of their life cycle completely.

This should close this issue: https://github.com/creativelifeform/three-nebula/issues/169

I also built a sandbox example to illustrate/test this, and it seems to work fine with both the gpu-renderer and the sprite-renderer.

rohan-deshpande commented:
There's a single test failing here, seems like it's related to an expectation that should now be different. If you could update the PR with a new test to cover your functionality that would be great, otherwise I will try to get around to it soon.

rohan-deshpande commented on sandbox/common/Visualization.js:
```diff
@@ -106,7 +106,7 @@ window.Visualization = class {
     } = this;
 
-    console.log(this);
+    //console.log(this);
```
```suggestion
```

rohan-deshpande commented on src/emitter/Emitter.js:
```diff
@@ -568,7 +569,11 @@ export default class Emitter extends Particle {
     }
 
-    this.generate(time);
+    if (this.isEmitting)
```
```suggestion
    if (this.isEmitting) {
```

rohan-deshpande replied:
Just stylistic stuff

```
if (this.isEmitting) {
  this.generate(time);
}
```

rohan-deshpande commented on src/emitter/Emitter.js:
```diff
@@ -581,6 +586,10 @@ export default class Emitter extends Particle {
         this.bindEmitterEvent && this.dispatch(PARTICLE_DEAD, particle);
         this.parent.pool.expire(particle.reset());
         this.particles.splice(i, 1);
+        if(this.particles.length === 0)
```
More stylistic stuff

```
        if (this.particles.length === 0) {
          this.parent && this.parent.dispatch(SYSTEM_UPDATE);
        }
```

rohan-deshpande commented:
Not essential as it might be a bit challenging but it would be good if we could add a unit test here to ensure that particles do indeed live out their entire life. I might try to add it after this gets merged. LGTM if we can fix the broken test I'll merge it. Thanks so much for this, I knew it would be a simple fix but have just not had the time to get it done.

andrewleek-droplab commented:
@rohan-deshpande Okay, I have fixed that failing test and added a test to verify that particles live out their lives after stopEmit() is called.  Please take a look as this test stuff is a bit new to me.

rohan-deshpande commented:
Weird I thought I dropped support for Node 10.x that pipeline should be gone, I will double check but this LGTM!
